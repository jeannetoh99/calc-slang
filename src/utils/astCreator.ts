import * as es from '../ast'

export const getValueDeclarationName = (decl: es.ValueDeclaration) =>
  (decl.declarations[0].id as es.Identifier).name

export const locationDummyNode = (line: number, column: number) =>
  literal('Dummy', 'string', { start: { line, column }, end: { line, column } })

export const identifier = (name: string, loc?: es.SourceLocation | null): es.Identifier => ({
  type: 'Identifier',
  name,
  loc
})

export const literal = (
  value: any,
  litType: es.Type,
  loc?: es.SourceLocation | null
): es.Literal => ({
  type: 'Literal',
  value,
  litType,
  loc
})

export const expressionStatement = (expression: es.Expression): es.ExpressionStatement => ({
  type: 'ExpressionStatement',
  expression
})

export const program = (body: es.Statement[]): es.Program => ({
  type: 'Program',
  sourceType: 'module',
  body
})
