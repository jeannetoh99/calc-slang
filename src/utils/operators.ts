import { Thunk } from '../types'

export function forceIt(val: Thunk | any): any {
  if (val !== undefined && val !== null && val.isMemoized !== undefined) {
    if (val.isMemoized) {
      return val.memoizedValue
    }

    const evaluatedValue = forceIt(val.f())

    val.isMemoized = true
    val.memoizedValue = evaluatedValue

    return evaluatedValue
  } else {
    return val
  }
}
