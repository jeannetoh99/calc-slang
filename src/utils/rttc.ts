import * as es from '../ast'
import { RuntimeSourceError } from '../errors/runtimeSourceError'
import { ErrorSeverity, ErrorType } from '../types'

export class TypeError extends RuntimeSourceError {
  public type = ErrorType.RUNTIME
  public severity = ErrorSeverity.ERROR
  public location: es.SourceLocation

  constructor(node: es.Node, public side: string, public expected: string, public got: string) {
    super(node)
  }

  public explain() {
    return `Expected ${this.expected}${this.side}, got ${this.got}.`
  }

  public elaborate() {
    return this.explain()
  }
}

export const isInt = (v: es.Literal) => v.smlType.type === 'int'
export const isReal = (v: es.Literal) => v.smlType.type === 'real'
export const isNum = (v: es.Literal) => isInt(v) || isReal(v)
export const isString = (v: es.Literal) => v.smlType.type === 'string'
export const isBool = (v: es.Literal) => v.smlType.type === 'bool'
export const isUnit = (v: es.Literal) => v.smlType.type === 'unit'

export const checkIsInt = (node: es.Node, side: string, test: es.Literal) => {
  return isInt(test) ? undefined : new TypeError(node, side, 'int', test.smlType.type)
}

export const checkIsReal = (node: es.Node, side: string, test: es.Literal) => {
  return isReal(test) ? undefined : new TypeError(node, side, 'real', test.smlType.type)
}

export const checkIsNum = (node: es.Node, side: string, test: es.Literal) => {
  return isNum(test) ? undefined : new TypeError(node, side, 'real or int', test.smlType.type)
}

export const checkIsString = (node: es.Node, side: string, test: es.Literal) => {
  return isString(test) ? undefined : new TypeError(node, side, 'string', test.smlType.type)
}

export const checkIsBool = (node: es.Node, side: string, test: es.Literal) => {
  return isBool(test) ? undefined : new TypeError(node, side, 'boolean', test.smlType.type)
}

export const checkIsUnit = (node: es.Node, side: string, test: es.Literal) => {
  return isUnit(test) ? undefined : new TypeError(node, side, 'unit', test.smlType.type)
}

export const checkIsList = (node: es.Node, side: string, test: es.Literal) => {
  return isUnit(test) ? undefined : new TypeError(node, side, 'unit', test.smlType.type)
}


export const checkIsType = (node: es.Node, side: string, test: es.Literal, type: es.LiteralType) => {
  switch (type.type) {
    case 'int':
      return checkIsInt(node, side, test)
    case 'real':
      return checkIsReal(node, side, test)
    case 'string':
      return checkIsString(node, side, test)
    case 'bool':
      return checkIsBool(node, side, test)
    case 'unit':
      return checkIsUnit(node, side, test)
  }
}
