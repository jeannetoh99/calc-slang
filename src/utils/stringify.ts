import { isInteger } from 'lodash'

import { List, ListType, SmlValue, Tuple, TupleType, Type, VariableType } from '../ast'
import { ProgramResult, ResultType, Value } from '../types'

export interface ArrayLike {
  replPrefix: string
  replSuffix: string
  replArrayContents: () => Value[]
}

export type DecResType = {
  name: string
  value: Value
  type: string
}

const formatResult = (result: DecResType) => {
  return ['val', result.name, '=', result.value, ':', result.type].join(' ') + ';'
}

export const extractVariableTypes = (type: Type): VariableType[] => {
  switch (type.type) {
    case 'function':
      return extractVariableTypes(type.paramType).concat(extractVariableTypes(type.returnType))
    case 'list':
      return extractVariableTypes(type.elementType)
    case 'tuple':
      return type.elementTypes.flatMap(elemType => extractVariableTypes(elemType))
    case 'variable':
      return [type]
    default:
      return []
  }
}

export const stringifyVarTypes = (type: Type): string => {
  const varTypes = extractVariableTypes(type)
  let res = ''
  if (varTypes.length != 0) {
    res = '\u2200 ' + stringifyType(varTypes[0])
    for (let i = 1; i < varTypes.length; i++) {
      res += ',' + stringifyType(varTypes[i])
    }
    res += ' . '
  }
  return res
}

export const stringifyType = (type: Type, inner: boolean = false): string => {
  return type.type === 'function'
    ? (type.paramType ? stringifyType(type.paramType) : '?') +
        ' -> ' +
        (type.returnType ? stringifyType(type.returnType) : '?')
    : type.type === 'list'
    ? (type.elementType ? stringifyType(type.elementType) : '?') + ' list'
    : type.type === 'tuple'
    ? (inner ? '(' : '') +
      type.elementTypes.map(elemType => stringifyType(elemType, true)).join(' * ') +
      (inner ? ')' : '')
    : type.type === 'variable'
    ? "'v" + type.id
    : type.type
}

const stringifyValue = (value: SmlValue, type: Type): Value => {
  switch (type.type) {
    case 'string': {
      return '"' + value.value + '"'
    }
    case 'unit': {
      return '()'
    }
    case 'real': {
      return value.value + (isInteger(value.value) ? '.0' : '')
    }
    case 'list': {
      const list = value as List
      return '[' + list.value.map(val => stringifyValue(val, type.elementType)) + ']'
    }
    case 'tuple': {
      const tuple = value as Tuple
      const elementStrings = []
      for (let i = 0; i < tuple.value.length; i++) {
        elementStrings.push(stringifyValue(tuple.value[i], type.elementTypes[i]))
      }
      return '(' + elementStrings + ')'
    }
    default: {
      return value.value
    }
  }
}

const extractMatch = (name: string, value: Value, type: Type): DecResType => {
  return {
    name,
    value: stringifyValue(value, type),
    type: stringifyVarTypes(type) + stringifyType(type)
  }
}

const extractDeclaration = (res: ResultType, type: Type): DecResType[] => {
  console.log(res, type)
  if (res.pat.type === 'Identifier') {
    return [extractMatch(res.pat.name, res.value, type)]
  }

  if (res.pat.type === 'TuplePattern') {
    const matches: DecResType[] = []
    const elemTypes = (type as TupleType).elementTypes
    const patterns = res.pat.elements
    const values = (res.value as Tuple).value
    for (let i = 0; i < res.pat.elements.length; i++) {
      const pat = patterns[i]
      if (pat.type === 'Identifier') {
        matches.push(extractMatch(pat.name, values[i], elemTypes[i]))
      }
    }
    return matches
  }

  // If pattern is wildcard or literal
  return []
}

export const formatResults = (result: DecResType[]) => {
  return result.map(formatResult).join('\n')
}

export const stringify = (value: Value): string => {
  const programResult = value as ProgramResult
  const res = []
  for (let i = 0; i < programResult.values.length; i++) {
    res.push(...extractDeclaration(programResult.values[i], programResult.types[i]))
  }
  return res.map(formatResult).join('\n')
}
