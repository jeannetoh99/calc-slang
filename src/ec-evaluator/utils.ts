import { uniqueId } from 'lodash'

import { Context } from '..'
import * as es from '../ast'
import * as errors from '../errors/errors'
import { RuntimeSourceError } from '../errors/runtimeSourceError'
import { arity } from '../stdlib/misc'
import { Environment, Frame, Value } from '../types'
import * as ast from '../utils/astCreator'
import { builtinFunctions, builtinInfixFunctions } from './builtin'
import * as instr from './instrCreator'
import { Agenda } from './interpreter'
import { AgendaItem, BuiltinInstr, ClosureInstr, Instr, InstrType } from './types'

/**
 * Stack is implemented for agenda and stash registers.
 */
interface IStack<T> {
  push(...items: T[]): void
  pop(): T | undefined
  peek(): T | undefined
  size(): number
}

export class Stack<T> implements IStack<T> {
  // Bottom of the array is at index 0
  private storage: T[] = []

  public constructor() {}

  public push(...items: T[]): void {
    for (const item of items) {
      this.storage.push(item)
    }
  }

  public pop(): T | undefined {
    return this.storage.pop()
  }

  public peek(): T | undefined {
    if (this.size() === 0) {
      return undefined
    }
    return this.storage[this.size() - 1]
  }

  public size(): number {
    return this.storage.length
  }
}

/**
 * Typeguard for esNode to distinguish between program statements and instructions.
 *
 * @param command An AgendaItem
 * @returns true if the AgendaItem is an esNode and false if it is an instruction.
 */
export const isNode = (command: AgendaItem): command is es.Node => {
  return (command as es.Node).type !== undefined
}

/**
 * Typeguard for esIdentifier. To verify if an esNode is an esIdentifier.
 *
 * @param node an esNode
 * @returns true if node is an esIdentifier, false otherwise.
 */
export const isIdentifier = (node: es.Node): node is es.Identifier => {
  return (node as es.Identifier).name !== undefined
}

/**
 * A helper function for handling sequences of statements.
 * Statements must be pushed in reverse order, and each statement is separated by a pop
 * instruction so that only the result of the last statement remains on stash.
 * Value producing statements have an extra pop instruction.
 *
 * @param seq Array of statements.
 * @returns Array of commands to be pushed into agenda.
 */
export const handleSequence = (seq: es.Statement[]): AgendaItem[] => {
  const result: AgendaItem[] = []
  let valueProduced = false
  for (const command of seq) {
    valueProduced ? result.push(instr.popInstr()) : (valueProduced = true)
    result.push(command)
  }
  return result.reverse()
}

/**
 * Environments
 */

export const currentEnvironment = (context: Context) => context.runtime.environments[0]

export const createEnvironment = (
  closure: ClosureInstr,
  args: Value[],
  callExpression: es.CallExpression
): Environment => {
  const id = uniqueId()
  const environment: Environment = {
    name: isIdentifier(callExpression.callee)
      ? callExpression.callee.name
      : 'anonymous closure ' + id,
    tail: closure.env,
    head: {},
    id,
    callExpression: {
      ...callExpression,
      args
    }
  }
  closure.srcNode.params.forEach((param, index) => {
    environment.head[(param as es.Identifier).name] = args[index]
  })
  return environment
}

export const popEnvironment = (context: Context) => context.runtime.environments.shift()

export const pushEnvironment = (context: Context, environment: Environment) => {
  context.runtime.environments.unshift(environment)
  context.runtime.environmentTree.insert(environment)
}

export const createBlockEnvironment = (
  context: Context,
  name = 'blockEnvironment',
  head: Frame = {}
): Environment => {
  return {
    name,
    tail: currentEnvironment(context),
    head,
    id: uniqueId()
  }
}
/**
 * Variables
 */

const DECLARED_BUT_NOT_YET_ASSIGNED = Symbol('Used to implement hoisting')

function declareIdentifier(context: Context, name: string, node: es.Node) {
  const environment = currentEnvironment(context)
  if (!environment.head.hasOwnProperty(name)) {
    environment.head[name] = DECLARED_BUT_NOT_YET_ASSIGNED
  }
  return environment
}

function declareVariables(context: Context, node: es.ValueDeclaration) {
  for (const declaration of node.declarations) {
    declareIdentifier(context, (declaration.id as es.Identifier).name, node)
  }
}

export function declareFunctionsAndVariables(context: Context, node: es.BlockStatement) {
  for (const statement of node.body) {
    switch (statement.type) {
      case 'ValueDeclaration':
        declareVariables(context, statement)
        break
      case 'FunctionDeclaration':
        declareIdentifier(context, (statement.id as es.Identifier).name, statement)
        break
    }
  }
}

export function defineVariable(context: Context, name: string, value: Value, constant = false) {
  const environment = currentEnvironment(context)

  Object.defineProperty(environment.head, name, {
    value,
    writable: !constant,
    enumerable: true
  })

  return environment
}

export const getVariable = (context: Context, name: string, node: es.Identifier) => {
  let environment: Environment | null = currentEnvironment(context)
  while (environment) {
    if (environment.head.hasOwnProperty(name)) {
      if (environment.head[name] === DECLARED_BUT_NOT_YET_ASSIGNED) {
        return handleRuntimeError(context, new errors.UnassignedVariable(name, node))
      } else {
        return environment.head[name]
      }
    } else {
      environment = environment.tail
    }
  }
  return handleRuntimeError(context, new errors.UndefinedVariable(name, node))
}

export const handleRuntimeError = (context: Context, error: RuntimeSourceError) => {
  context.errors.push(error)
  context.runtime.environments = context.runtime.environments.slice(
    -context.numberOfOuterEnvironments
  )
  throw error
}

export const checkNumberOfArguments = (
  context: Context,
  callee: ClosureInstr | BuiltinInstr,
  args: Value[],
  exp: es.CallExpression
) => {
  if (callee?.instrType === InstrType.CLOSURE) {
    // User-defined or Pre-defined functions
    const instr = callee as ClosureInstr
    const params = instr.srcNode?.params
    if (params.length !== args.length) {
      return handleRuntimeError(
        context,
        new errors.InvalidNumberOfArguments(exp, params.length, args.length)
      )
    }
  } else if (callee?.instrType === InstrType.BUILTIN) {
    const instr = callee as BuiltinInstr
    if (instr.arity !== args.length) {
      return handleRuntimeError(
        context,
        new errors.InvalidNumberOfArguments(exp, instr.arity, args.length)
      )
    }
  }
  return undefined
}

/**
 * This function can be used to check for a stack overflow.
 * The current limit is set to be an agenda size of 1.0 x 10^5, if the agenda
 * flows beyond this limit an error is thrown.
 * This corresponds to about 10mb of space according to tests ran.
 */
export const checkStackOverFlow = (context: Context, agenda: Agenda) => {
  if (agenda.size() > 100000) {
    const stacks: es.CallExpression[] = []
    let counter = 0
    for (
      let i = 0;
      counter < errors.MaximumStackLimitExceeded.MAX_CALLS_TO_SHOW &&
      i < context.runtime.environments.length;
      i++
    ) {
      if (context.runtime.environments[i].callExpression) {
        stacks.unshift(context.runtime.environments[i].callExpression!)
        counter++
      }
    }
    handleRuntimeError(
      context,
      new errors.MaximumStackLimitExceeded(context.runtime.nodes[0], stacks)
    )
  }
}

export const populateBuiltInIdentifiers = (context: Context) => {
  for (const key in builtinFunctions) {
    const builtinInstr = instr.builtinInstr(key, arity(builtinFunctions[key]), false)
    defineVariable(context, key, builtinInstr)
  }
}
