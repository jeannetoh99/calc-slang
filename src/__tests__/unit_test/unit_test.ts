import { DecResType, formatResults } from '../../utils/stringify'
import { expectResult } from '../../utils/testing'

///////////////////// LITERALS /////////////////////

test('Empty code returns undefined', () => {
  const expected: DecResType[] = []
  return expectResult('').toBe(formatResults(expected))
})

test('Single integer self-evaluates to itself', () => {
  const expected: DecResType[] = [
    {
      name: 'it',
      value: 42,
      type: 'int'
    }
  ]
  return expectResult('42;').toBe(formatResults(expected))
})

test('Single boolean self-evaluates to itself', () => {
  const expected: DecResType[] = [
    {
      name: 'it',
      value: true,
      type: 'bool'
    }
  ]
  return expectResult('true;').toBe(formatResults(expected))
})

test('Single real self-evaluates to itself', () => {
  const expected: DecResType[] = [
    {
      name: 'it',
      value: 42.5,
      type: 'real'
    }
  ]
  return expectResult('42.5;').toBe(formatResults(expected))
})

test('Single string self-evaluates to itself', () => {
  const expected: DecResType[] = [
    {
      name: 'it',
      value: '"42"',
      type: 'string'
    }
  ]
  return expectResult('"42";').toBe(formatResults(expected))
})

test('Multiline string self-evaluates to itself', () => {
  const expected: DecResType[] = [
    {
      name: 'it',
      value: '"1\n1"',
      type: 'string'
    }
  ]
  return expectResult(`"1\\n1";`).toBe(formatResults(expected))
})

test('Unit evaluates to undefined', () => {
  const expected: DecResType[] = [
    {
      name: 'it',
      value: '()',
      type: 'unit'
    }
  ]
  return expectResult('();').toBe(formatResults(expected))
})

///////////////// LIST_EXPRESSION //////////////////

test('Integer list evaluates to list', () => {
  const expected: DecResType[] = [
    {
      name: 'it',
      value: '[1,2,3]',
      type: 'int list'
    }
  ]
  return expectResult('[1, 2, 3];').toBe(formatResults(expected))
})

test('Boolean list evaluates to list', () => {
  const expected: DecResType[] = [
    {
      name: 'it',
      value: '[true,false]',
      type: 'bool list'
    }
  ]
  return expectResult('[true, false];').toBe(formatResults(expected))
})

test('Real list evaluates to list', () => {
  const expected: DecResType[] = [
    {
      name: 'it',
      value: '[1.5,2.5,3.5]',
      type: 'real list'
    }
  ]
  return expectResult('[1.5, 2.5, 3.5];').toBe(formatResults(expected))
})

test('String list evaluates to list', () => {
  const expected: DecResType[] = [
    {
      name: 'it',
      value: '["1","2","3"]',
      type: 'string list'
    }
  ]
  return expectResult('["1", "2", "3"];').toBe(formatResults(expected))
})

test('List of lists evaluates to list', () => {
  const expected: DecResType[] = [
    {
      name: 'it',
      value: '[[1,2],[3,4]]',
      type: 'int list list'
    }
  ]
  return expectResult('[[1, 2], [3, 4]];').toBe(formatResults(expected))
})

test(`List of expressions evaluates to list`, () => {
  const expected: DecResType[] = [
    {
      name: 'it',
      value: '[3,7]',
      type: 'int list'
    }
  ]
  return expectResult('[1 + 2, 3 + 4];').toBe(formatResults(expected))
})

///////////////// CONDITIONAL_EXPRESSION //////////////////
test('If expression evaluates to then branch', () => {
  const expected: DecResType[] = [
    {
      name: 'it',
      value: '"a"',
      type: 'string'
    }
  ]
  return expectResult('if true then "a" else "b";').toBe(formatResults(expected))
})

///////////////// LAMBDA_EXPRESSION //////////////////
test('Lambda expression evaluates to function', () => {
  const expected: DecResType[] = [
    {
      name: 'it',
      value: 'fn',
      type: 'int -> int'
    }
  ]
  return expectResult('fn x => x + 1;').toBe(formatResults(expected))
})

///////////////// VALUE_DECLARATION //////////////////
test('Value declaration evaluates to value', () => {
  const expected: DecResType[] = [
    {
      name: 'x',
      value: '1',
      type: 'int'
    }
  ]
  return expectResult('val x = 1;').toBe(formatResults(expected))
})

///////////////// APPLICATION_EXPRESSION //////////////////
test('Application expression evaluates to function application', () => {
  const expected: DecResType[] = [
    {
      name: 'n',
      value: 'fn',
      type: 'int -> int'
    },
    {
      name: 'it',
      value: '2',
      type: 'int'
    }
  ]
  return expectResult('val n = fn x => x + 1; \n n 1;').toBe(formatResults(expected))
})

///////////////// EXPRESSION_STATEMENT //////////////////
test('Expression statement evaluates to unit', () => {
  const expected: DecResType[] = [
    {
      name: 'it',
      value: '()',
      type: 'unit'
    }
  ]
  return expectResult('();').toBe(formatResults(expected))
})

///////////////// TUPLE //////////////////
test('Tuple evaluates to tuple', () => {
  const expected: DecResType[] = [
    {
      name: 'it',
      value: '("a",1,1.0)',
      type: 'string * int * real'
    }
  ]
  return expectResult('("a", 1, 1.0);').toBe(formatResults(expected))
})

///////////////// PATTERN_MATCHING //////////////////
test('Pattern matching evaluates evaluates pairwise values', () => {
  const expected: DecResType[] = [
    {
      name: 'a',
      value: '2',
      type: 'int'
    },
    {
      name: 'b',
      value: '"s"',
      type: 'string'
    },
    {
      name: 'c',
      value: '3.4',
      type: 'real'
    },
    {
      name: 'd',
      value: '[1,2]',
      type: 'int list'
    }
  ]
  return expectResult('val (a,b,c,d) = (2, "s", 3.4, [1,2]);').toBe(formatResults(expected))
})

///////////////// FUNCTION_DECLARATION //////////////////
test('Function declaration evaluates to function', () => {
  const expected: DecResType[] = [
    {
      name: 'it',
      value: 'fn',
      type: 'int -> int'
    }
  ]
  return expectResult('fn x => x+1;').toBe(formatResults(expected))
})

///////////////// RECURSIVE_VALUE_DECLARATION //////////////////
test('Recursive value declaration', () => {
  const expected: DecResType[] = [
    {
      name: 'fact',
      value: 'fn',
      type: 'int -> int'
    }
  ]
  return expectResult('val rec fact = fn n => if n=0 then 1 else n*fact(n-1);').toBe(
    formatResults(expected)
  )
})

///////////////// TYPE_DECLARATION //////////////////
test('Type declaration evaluates to type', () => {
  const expected: DecResType[] = [
    {
      name: 'it',
      value: '100',
      type: 'int'
    }
  ]
  return expectResult('100: int;').toBe(formatResults(expected))
})

///////////////// SEQUENCE_EXPRESSION //////////////////
test('Sequence expression evaluates to last expression', () => {
  const expected: DecResType[] = [
    {
      name: 'it',
      value: '"ab"',
      type: 'string'
    }
  ]
  return expectResult('( 1 + 2; "a"^"b" );').toBe(formatResults(expected))
})

///////////////// LET_EXPRESSION //////////////////
test('Let expression', () => {
  const expected: DecResType[] = [
    {
      name: 'it',
      value: '0',
      type: 'int'
    }
  ]
  return expectResult('let val x=0 in if x=0 then 0 else 100 div x end;').toBe(
    formatResults(expected)
  )
})
