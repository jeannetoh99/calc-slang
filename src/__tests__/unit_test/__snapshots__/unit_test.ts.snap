// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application expression evaluates to function application: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "val n = fn x => x + 1; 
 n 1;",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "val n = fn : int -> int;
val it = 2 : int;",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Boolean list evaluates to list: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "[true, false];",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "val it = [true,false] : bool list;",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Empty code returns undefined: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Expression statement evaluates to unit: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "();",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "val it = () : unit;",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Function declaration evaluates to function: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "fn x => x+1;",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "val it = fn : int -> int;",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`If expression evaluates to then branch: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "if true then \\"a\\" else \\"b\\";",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "val it = \\"a\\" : string;",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Integer list evaluates to list: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "[1, 2, 3];",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "val it = [1,2,3] : int list;",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Lambda expression evaluates to function: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "fn x => x + 1;",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "val it = fn : int -> int;",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`List of expressions evaluates to list: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "[1 + 2, 3 + 4];",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "val it = [3,7] : int list;",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`List of lists evaluates to list: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "[[1, 2], [3, 4]];",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "val it = [[1,2],[3,4]] : int list list;",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Multiline string self-evaluates to itself: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "\\"1\\\\n1\\";",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "val it = \\"1
1\\" : string;",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Pattern matching evaluates evaluates pairwise values: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "val (a,b,c,d) = (2, \\"s\\", 3.4, [1,2]);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "val a = 2 : int;
val b = \\"s\\" : string;
val c = 3.4 : real;
val d = [1,2] : int list;",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Real list evaluates to list: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "[1.5, 2.5, 3.5];",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "val it = [1.5,2.5,3.5] : real list;",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Recursive value declaration: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "val rec fact = fn n => if n=0 then 1 else n*fact(n-1);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "val fact = fn : int -> int;",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Single boolean self-evaluates to itself: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "true;",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "val it = true : bool;",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Single integer self-evaluates to itself: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "42;",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "val it = 42 : int;",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Single real self-evaluates to itself: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "42.5;",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "val it = 42.5 : real;",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Single string self-evaluates to itself: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "\\"42\\";",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "val it = \\"42\\" : string;",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`String list evaluates to list: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "[\\"1\\", \\"2\\", \\"3\\"];",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "val it = [\\"1\\",\\"2\\",\\"3\\"] : string list;",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Tuple evaluates to tuple: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "(\\"a\\", 1, 1.0);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "val it = (\\"a\\",1,1.0) : string * int * real;",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Type declaration evaluates to type: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "100: int;",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "val it = 100 : int;",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Unit evaluates to undefined: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "();",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "val it = () : unit;",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Value declaration evaluates to value: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "val x = 1;",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "val x = 1 : int;",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;
